tests/Installer/PackagesInstallerTest.php

```diff
@@ -11,121 +11,131 @@
             $typeResolver,
             $io,
             $depInstaller
         );
 
-        if ($expected) {
-            $depInstaller
-                ->expects(self::exactly(count($expected)))
-                ->method('installPackage')
-                ->withConsecutive(...$expected);
-        } else {
-            $depInstaller
-                ->expects(self::never())
-                ->method('installPackage');
-        }
+        $depInstaller
+            ->expects($this->atLeastOnce())
+            ->method('installPackage');
 
         $installer->install();
     }
 
     /**
-     * @return array
+     * @throws Exception
      */
-    public function dataProvider(): array
+    public function testCanMergeDefaultPackagesWhenInstalling(): void
     {
-        return [
-            [
-                'magento1',
-                $this->createLinkMocks(['foo/bar']),
-                [['youwe/coding-standard-magento1']]
-            ],
-            [
-                'magento1',
-                $this->createLinkMocks(
-                    ['foo/bar', 'youwe/coding-standard-magento1']
-                ),
-                [['youwe/coding-standard-magento1']]
-            ],
-            [
-                'magento2',
-                $this->createLinkMocks(['foo/bar']),
-                [['youwe/coding-standard-magento2']]
+        $composer     = $this->createMock(Composer::class);
+        $typeResolver = $this->createMock(ProjectTypeResolver::class);
+        $depInstaller = $this->createMock(DependencyInstaller::class);
+        $io           = $this->createMock(IOInterface::class);
+
+        // Simulate magento 1 project
+        $typeResolver
+            ->method('resolve')
+            ->willReturn('magento1');
+
+        $mapping = [
+            MappingResolver::DEFAULT_MAPPING_TYPE => [
+                [
+                    'name' => 'phpunit/phpunit',
+                    'version' => '@stable',
+                    'dev' => true,
+                    'allowVersionOverride' => false
+                ]
             ],
-            [
-                'default',
-                $this->createLinkMocks(['foo/bar']),
-                null
+            'magento1' => [
+                [
+                    'name' => 'youwe/coding-standard-magento1',
+                    'version' => '^1.3.0',
+                    'dev' => true,
+                    'allowVersionOverride' => true
+                ]
             ],
-            [
-                'unknown',
-                $this->createLinkMocks(['foo/bar']),
-                null
-            ]
         ];
+
+        $installer = new PackagesInstaller(
+            $composer,
+            $typeResolver,
+            $io,
+            $depInstaller,
+            $mapping
+        );
+
+        // We expect 2 packages to be installed due to the mapping merge
+        $depInstaller
+            ->expects($this->exactly(2))
+            ->method('installPackage');
```

Comment:
FIXME: You're not checking the arguments provided to the `installPackage` but these are actually relevant. Let me know if you need help with consecutive calls argument checking

Reply:
Added this to a seperate test case to check for the recursive array merge, since I would expect a project-type related package to merge into a default package when the name conflicts. This way, project-type specific packages expected to be installed upstream can overwrite the defaults if needed too, and this merge is also unit tested seperately

Reply:
Still relevant here that you verify that both packages are installed  -- that the correct package is installed is more relevant then that the correct information message is written to io ;)

Reply:
I'll have a second look. If only withConsecutive() was not removed or you could chain with()'s on the same method :upside_down_face:

Reply:
Resolved through a callback checking the params manually. Also merged due to approval, if you disagree with the implementation feel free to raise it through me or the DevOps board so I can include it other upcoming changes.

Reply:
LGTM!